File: JavaLibrary/src/main/java/com/flat/simplex/parser/logic/block/BlockFor.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.Parser;
import com.flat.simplex.parser.logic.Block;
import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.error.Error;

import java.util.ArrayList;

public class BlockFor extends Block {

    private Token tokenInit;
    private Token tokenInitEnd;
    private Token tokenCondition;
    private Token tokenConditionEnd;
    private Token tokenLoop;
    private Token tokenLoopEnd;
    private Token tokenContent;
    private Token tokenContentEnd;
    private boolean commandBlock;

    private ArrayList<Block> blocks;
    private Block initLine;
    private BlockLine conditionLine;
    private BlockLine loopLine;

    public BlockFor(Context context, Block parent, Token start, Token end) {
        super(context, parent, start);

        Token token = start;
        Token lToken = start;
        int state = 0;
        while (token != end && token != null) {
            if (state == 0 && token.getKey() == Key.For) {
                state = 1;
            } else if (state == 1 && token.getKey() == Key.Param) {
                state = 2;
                readTriple(token);
            } else if (state == 2 && token.getKey() == Key.Brace) {
                state = 3;
                tokenContent = token;
                commandBlock = true;
            } else if (state == 2) {
                state = 3;
                tokenContent = token;
                tokenContentEnd = end;
                break;
            } else {
                context.error(token, Error.unexpectedToken);
            }
            lToken = token;
            token = token.getNext();
        }
        if (state < 3) {
            context.error(lToken, Error.unexpectedEndOfTokens);
        }
    }

    private void readTriple(Token param) {
        Token lToken = param;
        Token token = param.getChild();
        Token end = param.getLastChild();
        int state = 0;
        while (token != end && token != null) {
            if (state == 0 && token.getKey() != Key.Semicolon) {
                tokenInit = token;
                tokenInitEnd = token.getNext();
                state = 1;
            } else if (state == 0 && token.getKey() == Key.Semicolon) {
                state = 2;
            } else if (state == 1 && token.getKey() != Key.Semicolon) {
                tokenInitEnd = token.getNext();
            } else if (state == 1 && token.getKey() == Key.Semicolon) {
                state = 2;
            } else if (state == 2 && token.getKey() != Key.Semicolon) {
                tokenCondition = token;
                tokenConditionEnd = token.getNext();
                state = 3;
            } else if (state == 2 && token.getKey() == Key.Semicolon) {
                state = 4;
            } else if (state == 3 && token.getKey() != Key.Semicolon) {
                tokenConditionEnd = token.getNext();
            } else if (state == 3 && token.getKey() == Key.Semicolon) {
                state = 4;
            } else if (state == 4 && token.getKey() != Key.Semicolon) {
                tokenLoop = token;
                tokenLoopEnd = token.getNext();
                state = 5;
            } else if (state == 5 && token.getKey() != Key.Semicolon) {
                tokenLoopEnd = token.getNext();
            } else {
                if (state >= 4) state = 6;
                getContext().error(token, Error.unexpectedToken);
            }
            lToken = token;
            token = token.getNext();
        }
        if (state < 4) {
            getContext().error(lToken, Error.unexpectedEndOfTokens);
        }
    }

    @Override
    public void read() {
        if (tokenContent != null) {
            if (commandBlock) {
                blocks = new Parser(getContext(), this).parse(tokenContent.getChild(), tokenContent.getLastChild());
                if (tokenContent.getLastChild() == null) {
                    getContext().error(tokenContent, Error.missingCloser);
                }
            } else {
                blocks = new Parser(getContext(), this).parse(tokenContent, tokenContentEnd);
            }
        }
        if (tokenInit != null && tokenInit != tokenInitEnd) {
            if (tokenInit.getKey() == Key.Var) {
                initLine = new BlockVar(getContext(), this, tokenInit, tokenInitEnd, false);
            } else {
                initLine = new BlockLine(getContext(), this, tokenInit, tokenInitEnd, false);
            }
            initLine.read();
        }
        if (tokenCondition != null && tokenCondition != tokenConditionEnd) {
            conditionLine = new BlockLine(getContext(), this, tokenCondition, tokenConditionEnd, false);
            conditionLine.read();
        }
        if (tokenLoop != null && tokenLoop != tokenLoopEnd) {
            loopLine = new BlockLine(getContext(), this, tokenLoop, tokenLoopEnd, false);
            loopLine.read();
        }
    }

    @Override
    public boolean isLoop() {
        return true;
    }

    public Token getTokenInit() {
        return tokenInit;
    }

    public Token getTokenInitEnd() {
        return tokenInitEnd;
    }

    public Token getTokenCondition() {
        return tokenCondition;
    }

    public Token getTokenConditionEnd() {
        return tokenConditionEnd;
    }

    public Token getTokenLoop() {
        return tokenLoop;
    }

    public Token getTokenLoopEnd() {
        return tokenLoopEnd;
    }

    public Token getTokenContent() {
        return tokenContent;
    }

    public Token getTokenContentEnd() {
        return tokenContentEnd;
    }

    public boolean isCommandBlock() {
        return commandBlock;
    }

    public ArrayList<Block> getBlocks() {
        return blocks;
    }
}

File: SimplexApp/simplex/parser/logic/block/BlockFor.js
var Key, BlockLine, BlockVar, Parser, Error;
const Block = require("simplex/parser/logic/Block.js");

class BlockFor extends Block {
    constructor(parent, start, end) {
        super(parent, start);

        this.tokenInit = null;
        this.tokenInitEnd = null;
        this.tokenCondition = null;
        this.tokenConditionEnd = null;
        this.tokenLoop = null;
        this.tokenLoopEnd = null;
        this.tokenContent = null;
        this.tokenContentEnd = null;
        this.commandBlock = false;
        this.blocks = null;
        this.initLine = null;
        this.conditionLine = null;
        this.loopLine = null;

        let token = start;
        let lToken = start;
        let state = 0;
        while (token !== end && token !== null) {
            if (state === 0 && token.getKey() === Key.For) {
                state = 1;
            } else if (state === 1 && token.getKey() === Key.Param) {
                state = 2;
                this.readTriple(token);
            } else if (state === 2 && token.getKey() === Key.Brace) {
                state = 3;
                this.tokenContent = token;
                this.commandBlock = true;
            } else if (state === 2) {
                state = 3;
                this.tokenContent = token;
                this.tokenContentEnd = end;
                break;
            } else {
                this.error(token, Error.unexpectedToken);
            }
            lToken = token;
            token = token.getNext();
        }

        if (state < 3) {
            this.error(lToken, Error.unexpectedEndOfTokens);
        }
    }
    readTriple(param) {
        let lToken = param;
        let token = param.getChild();
        let end = param.getLastChild();
        let state = 0;
        while (token !== end && token !== null) {
            if (state === 0 && token.getKey() !== Key.Semicolon) {
                this.tokenInit = token;
                this.tokenInitEnd = token.getNext();
                state = 1;
            } else if (state === 0 && token.getKey() === Key.Semicolon) {
                state = 2;
            } else if (state === 1 && token.getKey() !== Key.Semicolon) {
                this.tokenInitEnd = token.getNext();
            } else if (state === 1 && token.getKey() === Key.Semicolon) {
                state = 2;
            } else if (state === 2 && token.getKey() !== Key.Semicolon) {
                this.tokenCondition = token;
                this.tokenConditionEnd = token.getNext();
                state = 3;
            } else if (state === 2 && token.getKey() === Key.Semicolon) {
                state = 4;
            } else if (state === 3 && token.getKey() !== Key.Semicolon) {
                this.tokenConditionEnd = token.getNext();
            } else if (state === 3 && token.getKey() === Key.Semicolon) {
                state = 4;
            } else if (state === 4 && token.getKey() !== Key.Semicolon) {
                this.tokenLoop = token;
                this.tokenLoopEnd = token.getNext();
                state = 5;
            } else if (state === 5 && token.getKey() !== Key.Semicolon) {
                this.tokenLoopEnd = token.getNext();
            } else {
                if (state >= 4) {
                    state = 6;
                }
                this.error(token, Error.unexpectedToken);
            }
            lToken = token;
            token = token.getNext();
        }

        if (state < 4) {
            this.error(lToken, Error.unexpectedEndOfTokens);
        }
    }
    read() {
        if (this.tokenInit !== null && this.tokenInit !== this.tokenInitEnd) {
            if (this.tokenInit.getKey() === Key.Var) {
                this.initLine = new BlockVar(this, this.tokenInit, this.tokenInitEnd, false);
            } else {
                this.initLine = new BlockLine(this, this.tokenInit, this.tokenInitEnd, false);
            }
            this.initLine.read();
        }
        if (this.tokenCondition !== null && this.tokenCondition !== this.tokenConditionEnd) {
            this.conditionLine = new BlockLine(this, this.tokenCondition, this.tokenConditionEnd, false);
            this.conditionLine.read();
        }
        if (this.tokenLoop !== null && this.tokenLoop !== this.tokenLoopEnd) {
            this.loopLine = new BlockLine(this, this.tokenLoop, this.tokenLoopEnd, false);
            this.loopLine.read();
        }
        if (this.tokenContent !== null) {
            if (this.commandBlock) {
                this.blocks = new Parser(this).parse(this.tokenContent.getChild(), this.tokenContent.getLastChild());
                if (this.tokenContent.getLastChild() === null) {
                    this.error(this.tokenContent, Error.missingCloser);
                }
            } else {
                this.blocks = new Parser(this).parse(this.tokenContent, this.tokenContentEnd);
            }
        }
    }

    isLoop() {
        return true;
    }

    getTokenInit() {
        return this.tokenInit;
    }

    getTokenInitEnd() {
        return this.tokenInitEnd;
    }

    getTokenCondition() {
        return this.tokenCondition;
    }

    getTokenConditionEnd() {
        return this.tokenConditionEnd;
    }

    getTokenLoop() {
        return this.tokenLoop;
    }

    getTokenLoopEnd() {
        return this.tokenLoopEnd;
    }

    getTokenContent() {
        return this.tokenContent;
    }

    getTokenContentEnd() {
        return this.tokenContentEnd;
    }

    isCommandBlock() {
        return this.commandBlock;
    }

    getBlocks() {
        return this.blocks;
    }
}

module.exports = BlockFor;
Key = require("simplex/lexer/Key.js");
BlockLine = require("simplex/parser/logic/block/BlockLine.js");
BlockVar = require("simplex/parser/logic/block/BlockVar.js");
Parser = require("simplex/parser/Parser.js");
Error = require("simplex/parser/logic/error/Error.js");